core/nurse_scheduling/solver_pulp.py

```
@@ -0,0 +1,487 @@
+            raise NotImplementedError(
+                f"PuLP squared equality expects a bounded variable or constant, got {type(source_var)}."
+            )
+        if source_var.lowBound is None or source_var.upBound is None:
+            raise ValueError("Cannot linearize square for unbounded variable.")
+
+        inferred_lb, inferred_ub = int(source_var.lowBound), int(source_var.upBound)
+        lb, ub = source_var_range
+        lb = int(lb)
+        ub = int(ub)
+        if lb > ub:
+            raise ValueError(f"Invalid source variable bounds for square: [{lb}, {ub}]")
```

Ruff (0.15.1)

[warning] Avoid specifying long messages outside the exception class

(TRY003)